volatility3/framework/plugins/linux/kmsg.py

```diff
@@ -0,0 +1,386 @@
+
+                for line in self.get_log_lines(msg):
+                    yield facility_txt, level_txt, timestamp, caller, line
+                for line in self.get_dict_lines(msg):
+                    yield facility_txt, level_txt, timestamp, caller, line
+
+                cur_idx += msg.len
+
+            if cur_idx == end_idx:
+                break
+
+
+class KmsgFiveTen(ABCKmsg):
+    """In 5.10 the kernel ringbuffer implementation changed.
+    Previously only one process should read /proc/kmsg and it is permanently
+    open and periodically read by the syslog daemon.
+    A high level structure 'printk_ringbuffer' was added to represent the printk
+    ringbuffer which actually contains two ringbuffers. The descriptor ring
+    'desc_ring' contains the records' metadata, text offsets and states.
+    The data block ring 'text_data_ring' contains the records' text strings.
+    A pointer to the high level structure is kept in the prb pointer which is
+    initialized to a static ringbuffer.
+        static struct printk_ringbuffer *prb = &printk_rb_static;
+    In SMP systems with more than 64 CPUs this ringbuffer size is dynamically
+    allocated according the number of CPUs based on the value of
+    CONFIG_LOG_CPU_MAX_BUF_SHIFT. The prb pointer is updated consequently to
+    this dynamic ringbuffer in setup_log_buf().
+        prb = &printk_rb_dynamic;
+    Behind scenes, log_buf is still used as external buffer.
+    When the static printk_ringbuffer struct is initialized, _DEFINE_PRINTKRB
+    sets text_data_ring.data pointer to the address in log_buf which points to
+    the static buffer __log_buff.
+    If a dynamic ringbuffer takes place, setup_log_buf() sets
+    text_data_ring.data of printk_rb_dynamic to the new allocated external
+    buffer via the prb_init function.
+    In that case, the original external static buffer in __log_buf and
+    printk_rb_static are unused.
+        ...
+        new_log_buf = memblock_alloc(new_log_buf_len, LOG_ALIGN);
+        prb_init(&printk_rb_dynamic, new_log_buf, ...);
+        log_buf = new_log_buf;
+        prb = &printk_rb_dynamic;
+        ...
+    See printk.c and printk_ringbuffer.c in kernel/printk/ folder for more
+    details.
+    """
+    @classmethod
+    def symtab_checks(cls, vmlinux) -> bool:
+        return vmlinux.has_symbol('prb')
+
+    def get_text_from_data_ring(self, text_data_ring, desc, info) -> str:
+        text_data_sz = text_data_ring.size_bits
+        text_data_mask = 1 << text_data_sz
+
+        begin = desc.text_blk_lpos.begin % text_data_mask
+        end = desc.text_blk_lpos.next % text_data_mask
+
+        # This record doesn't contain text
+        if begin & 1:
+            return ""
+
+        # This means a wrap-around to the beginning of the buffer
+        if begin > end:
+            begin = 0
+
+        # Each element in the ringbuffer is "ID + data".
+        # See prb_data_ring struct
+        desc_id_size = 8  # sizeof(long)
```

Comment:
if this is truly a long value then sizeof() will be 4 on 32bit systems.

You can do something like:

module.get_type("long").size

To grab it dynamically.

I see its used in two places, so setting the size to an instance variable in __init__ with the other variables will make it accessible everywhere.

Reply:
Yeah, that's true and needs to be fixed. I use module.get_type(TYPE).size in other part of the code, but the reason I ended up hardcoing the value is the following:
>>> vmlinux.get_type('printk_log').size
16
>>> vmlinux.get_type('int').size
4
>>> vmlinux.get_type('long').size
Traceback (most recent call last):
  File "/home/user/volexity/volatility3/volatility3/framework/symbols/__init__.py", line 139, in _weak_resolve
    return getattr(self._dict[table_name], get_function)(component_name)
  File "/home/user/volexity/volatility3/volatility3/framework/symbols/intermed.py", line 54, in _delegate_function
    return getattr(self._delegate, name)(*args, **kwargs)
  File "/home/user/volexity/volatility3/volatility3/framework/symbols/intermed.py", line 674, in get_type
    return self.natives.get_type(self.name + constants.BANG + type_name)
  File "/home/user/volexity/volatility3/volatility3/framework/symbols/native.py", line 74, in get_type
    _native_type, native_format = self._native_dictionary[type_name]
KeyError: 'long'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/user/volexity/volatility3/volatility3/framework/contexts/__init__.py", line 159, in wrapper
    return getattr(self._context.symbol_space, method)(name)
  File "/home/user/volexity/volatility3/volatility3/framework/symbols/__init__.py", line 181, in get_type
    self._resolved[type_name] = self._weak_resolve(SymbolType.TYPE, type_name)  # type: ignore
  File "/home/user/volexity/volatility3/volatility3/framework/symbols/__init__.py", line 141, in _weak_resolve
    raise exceptions.SymbolError(component_name, table_name,
volatility3.framework.exceptions.SymbolError: Type vmlinux1!long references missing Type/Symbol/Enum: 'long'

Reply:
oh ok, checking the json ISF I figured out why that's happening. We need to use "long int" instead of "long":
```
>>> vmlinux.get_type('long int').size
8
```

Reply:
cool, fixed. It was actually needed the long unsigned int type there.

Reply:
thanks @atcuno for the review ;)

Reply:
So, as you discovered, the names of base types aren't necessarily fixed.  This will probably work fine, but if it's an address you're after, you should always be able to look up `pointer` and that should have the right size for the architecture...